rubix-core/src/main/java/com/qubole/rubix/core/ReadRequestChain.java

```diff
@@ -46,19 +48,32 @@
   private static final Log log = LogFactory.getLog(ReadRequestChain.class);
 
   public ReadRequestChain(int generationNumber)
+  {
+    this(generationNumber, Long.MAX_VALUE);
+  }
+
+  // Caller responsible to keep maxReadRequestSize block aligned
+  public ReadRequestChain(int generationNumber, long maxReadRequestSize)
   {
     super();
     this.generationNumber = generationNumber;
+    this.maxReadRequestSize = maxReadRequestSize;
     this.threadName = Thread.currentThread().getName();
   }
 
+  @VisibleForTesting
+  public void setMaxReadRequestSize(long maxReadRequestSize)
+  {
+    this.maxReadRequestSize = maxReadRequestSize;
+  }
+
   // Should be added in forward seek fashion for better performance
   public void addReadRequest(ReadRequest readRequest)
   {
     checkState(!isLocked, "Adding request to a locked chain");
     log.debug(String.format("Request to add ReadRequest: [%d, %d, %d, %d, %d]", readRequest.getBackendReadStart(), readRequest.getBackendReadEnd(), readRequest.getActualReadStart(), readRequest.getActualReadEnd(), readRequest.getDestBufferOffset()));
     if (lastRequest == null) {
-      addRequest(readRequest);
+      lastRequest = readRequest;
```

Comment:
can you please explain this? with this change, we are adding the last request in the `lock()` method.

Reply:
Yes, earlier we used to keep the reference to lastRequest put on queue and collated new request with it if they were sequential. That was not clean if were to add the logic of splitting the request by size. Doing it this way was cleaner as the most recent readRequest goes into kind of a staging space and can be updated but ones gone into queue are immutable.